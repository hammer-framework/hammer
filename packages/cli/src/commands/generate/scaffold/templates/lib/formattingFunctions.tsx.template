import React from 'react'

import humanize from 'humanize-string'

const MAX_STRING_LENGTH = 150

export const formatEnum = (values: string | string[] | null | undefined) => {
  if (values) {
    if (Array.isArray(values)) {
      const humanizedValues = values.map((value) => humanize(value))
      return humanizedValues.join(', ')
    } else {
      return humanize(values as string)
    }
  }

  return ''
}

export const truncate = (value: string | number) => {
  const output = value?.toString()
  if (output?.length > MAX_STRING_LENGTH) {
    return output.substring(0, MAX_STRING_LENGTH) + '...'
  }
  return output ?? ''
}

export const jsonTruncate = (obj: unknown) => {
  return truncate(JSON.stringify(obj, null, 2))
}

export const timeTag = (datetime?: string) => {
  return datetime ? (
    <time dateTime={datetime} title={datetime}>
      {new Date(datetime).toUTCString()}
    </time>
  ) : (
    ''
  )
}

export const checkboxInputTag = (checked: boolean) => {
  return <input type="checkbox" checked={checked} disabled />
}
