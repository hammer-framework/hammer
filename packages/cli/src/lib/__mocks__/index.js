export const getPaths = () => MOCK_PATHS

export const MOCK_PATHS = {
  base: 'mock_basedir',
  generated: {
    base: '.mock_redwood',
    schema: '.mock_redwood/schema.graphql',
    types: {
      includes: '.mock_redwood/types/includes',
      mirror: '.mock_redwood/types/mirror',
    },
    prebuild: '.mock_redwood/prebuild',
  },
  scripts: 'mock_scripts_dir',
  api: {
    base: 'mock_api',
    dataMigrations: 'mock_dataMigrations',
    db: 'mock_db_dir',
    dbSchema: 'mock_db_schema_dir',
    functions: 'fmock_unctions_dir',
    graphql: 'mock_graphql_dir',
    lib: 'mock_lib_dir',
    generators: 'mock_generators_dir',
    config: 'mock_config_dir',
    services: 'mock_services_dir',
    directives: 'mock_directives_dir',
    src: 'mock_src_dir',
    dist: 'mock_api/dist',
    types: 'mock_api/types',
    models: 'mock_models_dir',
  },
  web: {
    routes: ['mock_mock/routes'],
    base: 'mock_web',
    pages: 'mock_pages_dir',
    components: 'mock_components_dir',
    layouts: 'mock_layouts_dir',
    src: 'mock_src_dir',
    generators: 'mock_generators_dir',
    app: 'mock_app_dir',
    index: 'mock_index_dir',
    config: 'mock_config_dir',
    webpack: 'mock_webpack_dir',
    postcss: 'mock_postcss_dir',
    storybookConfig: 'mock_storybookConfig_dir',
    storybookPreviewConfig: 'mock_storybookPreviewConfig_dir',
    storybookManagerConfig: 'mock_storybookManagerConfig_dir',
    dist: 'mock_web/dist',
    types: 'mock_web/types',
  },
}
