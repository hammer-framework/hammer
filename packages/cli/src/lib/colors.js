import chalk from 'chalk'

/**
 * To keep a consistent color/style palette between cli packages, such as
 * @redwood/cli and @redwood/create-redwood-app, please keep them compatible
 * with one and another. We'll might split up and refactor these into a
 * separate package when there is a strong motivation behind it.
 *
 * Current files:
 *
 * - packages/cli/src/lib/colors.js (this file)
 * - packages/create-redwood-app/src/create-redwood-app.js
 */
export default {
  error: chalk.bold.red,
  warning: chalk.keyword('orange'),
  success: chalk.greenBright,
  info: chalk.grey,

  bold: chalk.bold,
  underline: chalk.underline,

  green: chalk.green,

  header: chalk.bold.underline.hex('#e8e8e8'),
  cmd: chalk.hex('#808080'),
  redwood: chalk.hex('#ff845e'),
  love: chalk.redBright,
}
