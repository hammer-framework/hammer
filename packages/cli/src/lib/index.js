import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\``
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/Generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())
  return template({
    name,
    ...nameVariants(name),
    ...rest,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}

export * from './generateTempSchema'
