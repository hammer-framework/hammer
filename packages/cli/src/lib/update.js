import fs from 'fs'
import path from 'path'

import boxen from 'boxen'
import latestVersion from 'latest-version'
import semver from 'semver'

import { getConfig } from '@redwoodjs/internal/dist/config'

import { validateTag } from '../commands/upgrade'

import { setLock, unsetLock } from './locking'

import { getPaths } from './index'

/**
 * @const {number} The number of milliseconds between update checks (24 hours)
 */
const CHECK_PERIOD = 24 * 60 * 60_000

/**
 * @const {number} The number of milliseconds between showing a user an update notification (24 hours)
 */
const SHOW_PERIOD = 24 * 60 * 60_000

/**
 * @const {number} The default datetime for shownAt and checkedAt in milliseconds, corresponds to 2000-01-01T00:00:00.000Z
 */
export const DEFAULT_DATETIME_MS = 946684800000

/**
 * @const {string} The identifier used for the lock within the check function
 */
export const LOCK_IDENTIFIER = 'UPDATE-CHECK'

/**
 * Performs an update check to detect if a newer version of redwood is available and records the result to a file within .redwood for persistence
 */
export async function check() {
  try {
    setLock(LOCK_IDENTIFIER)

    // Read package.json and extract the @redwood/core version
    const packageJson = JSON.parse(
      fs.readFileSync(path.join(getPaths().base, 'package.json'))
    )
    let localVersion = packageJson.devDependencies['@redwoodjs/core']

    // Remove any leading non-digits, i.e. ^ or ~
    while (!/\d/.test(localVersion.charAt(0))) {
      localVersion = localVersion.substring(1)
    }

    // Determine if the user has a tag (e.g. -rc, -canary), if so extract the tag from the version
    let tag = ''
    if (localVersion.includes('-')) {
      tag = extractTagFromVersion(localVersion)

      // Validate the local version, just incase the user has manually fiddled with it
      // no if needed to check as it throws an error if not successful, will be caught by listr
      validateTag(tag)
    }

    // Fetch the latest version from npm registry
    let remoteVersion
    try {
      remoteVersion = await latestVersion(
        '@redwoodjs/core',
        tag ? { version: tag } : {}
      )
    } catch (e) {
      console.error(e)
      throw new Error(
        `Could not find the latest version${tag ? ` with tag: ${tag}` : ''}`
      )
    }

    // Save the latest update information
    updateUpdateFile({
      localVersion,
      remoteVersion,
      checkedAt: new Date().getTime(),
    })
  } finally {
    unsetLock(LOCK_IDENTIFIER)
  }
}

/**
 * Determines if background checks are enabled. Checks are enabled within the redwood.toml config or by the `REDWOOD_BACKGROUND_UPDATE_CHECKS_ENABLED` env var.
 * @return {boolean} True if background update checks are enabled
 */
export function isBackgroundCheckEnabled() {
  return (
    getConfig().background.updateChecks ||
    process.env.REDWOOD_BACKGROUND_UPDATE_CHECKS_ENABLED
  )
}

/**
 * Determines if an update check is due based on if enough time has elapsed since the last check
 * @return {boolean} True if an update check is overdue
 * @see {@link CHECK_PERIOD} for the time between notifications
 */
export function shouldCheck() {
  const data = readUpdateFile()
  return data.checkedAt < new Date().getTime() - CHECK_PERIOD
}

/**
 * Determines if the user should see an update notification based on if a new version is available and enough time has elapsed since the last notification
 * @return {boolean} True if the user should see an update notification
 * @see {@link SHOW_PERIOD} for the time between notifications
 */
export function shouldShow() {
  const data = readUpdateFile()
  return (
    data.shownAt < new Date().getTime() - SHOW_PERIOD &&
    semver.gt(data.remoteVersion, data.localVersion)
  )
}

/**
 * Prints the update notification message to the console and updates the stored shownAt property
 * @see {@link getUpgradeMessage} for the definition of the string which is printed
 */
export function showUpgradeMessage() {
  console.log(getUpgradeMessage())
  updateUpdateFile({ shownAt: new Date().getTime() })
}

/**
 * Returns a nicely formatted string containing an update notification
 * @return {string} A specifically formatted update notification message
 */
export function getUpgradeMessage() {
  const data = readUpdateFile()
  if (semver.gt(data.remoteVersion, data.localVersion)) {
    let message = `  Checklist:\n   1. Read the release notes at: https://github.com/redwoodjs/redwood/releases  \n   2. Run "yarn rw upgrade" to upgrade`
    return boxen(message, {
      padding: 0,
      margin: 1,
      title: `Redwood Upgrade Available: ${data.localVersion} -> ${data.remoteVersion}`,
      borderColor: `#0b8379`, // The RedwoodJS colour
      borderStyle: 'round',
    })
  }
  return boxen(
    `  You are currently using version ${data.localVersion} which is the latest redwood version.  `,
    {
      padding: 0,
      margin: 1,
      title: `No upgrade is available`,
      borderColor: `#0b8379`, // The RedwoodJS colour
      borderStyle: 'round',
    }
  )
}

/**
 * Reads update data from a file within .redwood
 * @return {Object} The update data object containing the localVersion, remoteVersion, checkedAt and shownAt properties
 */
export function readUpdateFile() {
  const updateFilePath = path.join(
    getPaths().generated.base,
    'update-data.json'
  )
  try {
    return JSON.parse(fs.readFileSync(updateFilePath))
  } catch (error) {
    // Return the default if no existing update file is found
    if (error.code === 'ENOENT') {
      return {
        localVersion: '0.0.0',
        remoteVersion: '0.0.0',
        checkedAt: DEFAULT_DATETIME_MS,
        shownAt: DEFAULT_DATETIME_MS,
      }
    }
    throw error
  }
}

/**
 * Writes update data to a file within .redwood for persistence
 * @param {Object} updateData The data to persist.
 * @param {string} updateData.localVersion The version of the users current redwood project
 * @param {string} updateData.remoteVersion The latest released version of redwood
 * @param {number} updateData.checkedAt The last time an update check was performed, in milliseconds
 * @param {number} updateData.shownAt The last time an update notification was shown to the user, in milliseconds
 */
function updateUpdateFile({
  localVersion,
  remoteVersion,
  checkedAt,
  shownAt,
} = {}) {
  const updateFilePath = path.join(
    getPaths().generated.base,
    'update-data.json'
  )
  const existingData = readUpdateFile()
  const isNewerUpgrade = semver.gt(
    remoteVersion ?? '0.0.0',
    existingData.remoteVersion
  )
  const updatedData = {
    localVersion: localVersion ?? existingData.localVersion,
    remoteVersion: remoteVersion ?? existingData.remoteVersion,
    checkedAt: checkedAt ?? existingData.checkedAt,
    shownAt:
      shownAt ?? (isNewerUpgrade ? DEFAULT_DATETIME_MS : existingData.shownAt), // We reset the shownAt if a newer version becomes available
  }
  fs.writeFileSync(updateFilePath, JSON.stringify(updatedData, null, 2))
}

function extractTagFromVersion(version) {
  let tag = version.substring(version.indexOf('-') + 1).trim()
  if (tag.includes('.')) {
    tag = tag.split('.')[0]
  }
  return tag
}
