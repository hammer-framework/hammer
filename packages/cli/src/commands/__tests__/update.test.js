global.__dirname = __dirname

jest.mock('fs')
jest.mock('latest-version')

jest.mock('@redwoodjs/internal/dist/paths', () => {
  return {
    getConfigPath: () => '/mocked/project/redwood.toml',
    getPaths: () => {
      return {
        generated: {
          base: '.redwood',
        },
        base: '',
      }
    },
  }
})

import fs from 'fs'

import Enquirer from 'enquirer'
import latestVersion from 'latest-version'

import { handler } from '../update'

import '../../lib/mockTelemetry'

const CURRENT_DATETIME = Date.parse('2022-01-02T12:00:00.000Z')
const DEFAULT_DATETIME = Date.parse('2000-01-02T12:00:00.000Z')
const DEFAULT_UPDATE_DATA = {
  localVersion: '0.0.0',
  remoteVersion: '0.0.0',
  skipVersion: '0.0.0',
  upgradeAvailable: false,
  lastChecked: DEFAULT_DATETIME,
  lastShown: DEFAULT_DATETIME,
}

describe('Upgrade is not available (1.0.0 -> 1.0.0)', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.setSystemTime(CURRENT_DATETIME)
    latestVersion.mockImplementation(() => {
      return '1.0.0'
    })
    fs.__setMockFiles({
      'package.json': JSON.stringify({
        devDependencies: {
          '@redwoodjs/core': '^1.0.0',
        },
      }),
      '.redwood/update-data.json': JSON.stringify(DEFAULT_UPDATE_DATA),
    })
    jest.spyOn(process.stdout, 'write').mockImplementation(() => true)
  })

  afterEach(() => {
    jest.useRealTimers()
    fs.__setMockFiles({})
    jest.restoreAllMocks()
  })

  test('Unskip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: DEFAULT_UPDATE_DATA.localVersion,
      remoteVersion: DEFAULT_UPDATE_DATA.remoteVersion,
      skipVersion: '0.0.0',
      upgradeAvailable: DEFAULT_UPDATE_DATA.upgradeAvailable,
      lastChecked: DEFAULT_UPDATE_DATA.lastChecked,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Unskip produces correct file', async () => {
    await handler({ listr2: { rendererSilent: true } })
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '1.0.0',
      skipVersion: '0.0.0',
      upgradeAvailable: false,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Skip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual(DEFAULT_UPDATE_DATA)
  })

  test('Skip produces correct file', async () => {
    await handler({ listr2: { rendererSilent: true } })
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '1.0.0',
      skipVersion: '1.0.0',
      upgradeAvailable: false,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME,
    })
  })

  test('Shows correct final message', async () => {
    await handler({})
    const lastListr2Output =
      process.stdout.write.mock.calls[
        process.stdout.write.mock.calls.length - 1
      ][0]
    expect(lastListr2Output).toMatchSnapshot()
  })

  test('No output when silent flag is set', async () => {
    await handler({ silent: true })
    expect(process.stdout.write.mock.calls.length).toBe(0)
  })

  test('Does not prompt for upgrade', async () => {
    let didPrompt = false
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      didPrompt = true
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    expect(didPrompt).toBe(false)
  })

  test('Creates correct update-data.json file', async () => {
    await handler({ listr2: { rendererSilent: true } })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '1.0.0',
      skipVersion: '0.0.0',
      upgradeAvailable: false,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME,
    })
  })
})

// ----------------------------------------------------------------------------------------

describe('Upgrade is available (1.0.0 -> 2.0.0)', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.setSystemTime(CURRENT_DATETIME)
    latestVersion.mockImplementation(() => {
      return '2.0.0'
    })
    fs.__setMockFiles({
      'package.json': JSON.stringify({
        devDependencies: {
          '@redwoodjs/core': '^1.0.0',
        },
      }),
      '.redwood/update-data.json': JSON.stringify(DEFAULT_UPDATE_DATA),
    })
    jest.spyOn(process.stdout, 'write')
  })

  afterEach(() => {
    jest.useRealTimers()
    fs.__setMockFiles({})
    jest.restoreAllMocks()
  })

  test('Unskip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: DEFAULT_UPDATE_DATA.localVersion,
      remoteVersion: DEFAULT_UPDATE_DATA.remoteVersion,
      skipVersion: '0.0.0',
      upgradeAvailable: DEFAULT_UPDATE_DATA.upgradeAvailable,
      lastChecked: DEFAULT_UPDATE_DATA.lastChecked,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Unskip produces correct file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '2.0.0',
      skipVersion: '0.0.0',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Skip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual(DEFAULT_UPDATE_DATA)
  })

  test('Skip produces correct file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '2.0.0',
      skipVersion: '2.0.0',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME, // current - minus 2 hours
    })
  })

  test('Shows correct final message', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({ enquirer: customEnquirer })
    const lastListr2Output =
      process.stdout.write.mock.calls[
        process.stdout.write.mock.calls.length - 1
      ][0]
    expect(lastListr2Output).toMatchSnapshot()
  })

  test('No output when silent flag is set', async () => {
    await handler({ silent: true })
    expect(process.stdout.write.mock.calls.length).toBe(0)
  })

  test('Does prompt for upgrade', async () => {
    let didPrompt = false
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      didPrompt = true
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    expect(didPrompt).toBe(true)
  })

  test('Creates correct update-data.json file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0',
      remoteVersion: '2.0.0',
      skipVersion: '0.0.0',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME,
    })
  })
})

// ----------------------------------------------------------------------------------------

describe('Upgrade is available (1.0.0-canary.100 -> 1.0.0-canary.200)', () => {
  beforeEach(() => {
    jest.useFakeTimers()
    jest.setSystemTime(CURRENT_DATETIME)
    latestVersion.mockImplementation(() => {
      return '1.0.0-canary.200'
    })
    fs.__setMockFiles({
      'package.json': JSON.stringify({
        devDependencies: {
          '@redwoodjs/core': '^1.0.0-canary.100',
        },
      }),
      '.redwood/update-data.json': JSON.stringify(DEFAULT_UPDATE_DATA),
    })
    jest.spyOn(process.stdout, 'write')
  })

  afterEach(() => {
    jest.useRealTimers()
    fs.__setMockFiles({})
    jest.restoreAllMocks()
  })

  test('Unskip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: DEFAULT_UPDATE_DATA.localVersion,
      remoteVersion: DEFAULT_UPDATE_DATA.remoteVersion,
      skipVersion: '0.0.0',
      upgradeAvailable: DEFAULT_UPDATE_DATA.upgradeAvailable,
      lastChecked: DEFAULT_UPDATE_DATA.lastChecked,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Unskip produces correct file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    await handler({ listr2: { rendererSilent: true }, unskip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0-canary.100',
      remoteVersion: '1.0.0-canary.200',
      skipVersion: '0.0.0',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME - 2 * 60 * 60_000, // current - minus 2 hours
    })
  })

  test('Skip produces correct file - no previous update', async () => {
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual(DEFAULT_UPDATE_DATA)
  })

  test('Skip produces correct file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    await handler({ listr2: { rendererSilent: true }, skip: true })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0-canary.100',
      remoteVersion: '1.0.0-canary.200',
      skipVersion: '1.0.0-canary.200',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME,
    })
  })

  test('Shows correct final message', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({ enquirer: customEnquirer })
    const lastListr2Output =
      process.stdout.write.mock.calls[
        process.stdout.write.mock.calls.length - 1
      ][0]
    expect(lastListr2Output).toMatchSnapshot()
  })

  test('No output when silent flag is set', async () => {
    await handler({ silent: true })
    expect(process.stdout.write.mock.calls.length).toBe(0)
  })

  test('Does prompt for upgrade', async () => {
    let didPrompt = false
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      didPrompt = true
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    expect(didPrompt).toBe(true)
  })

  test('Creates correct update-data.json file', async () => {
    const customEnquirer = new Enquirer({ show: false })
    customEnquirer.on('prompt', (prompt) => {
      prompt.value = false
      prompt.submit()
    })
    await handler({
      listr2: { rendererSilent: true },
      enquirer: customEnquirer,
    })
    const updateData = JSON.parse(fs.readFileSync('.redwood/update-data.json'))
    expect(updateData).toEqual({
      localVersion: '1.0.0-canary.100',
      remoteVersion: '1.0.0-canary.200',
      skipVersion: '0.0.0',
      upgradeAvailable: true,
      lastChecked: CURRENT_DATETIME,
      lastShown: CURRENT_DATETIME,
    })
  })
})
