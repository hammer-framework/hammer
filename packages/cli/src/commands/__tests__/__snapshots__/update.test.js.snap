// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upgrade is available (1.0.0 -> 2.0.0) Shows correct final message 1`] = `
"[32mâœ”[39m [38;2;11;131;121m[39m
  [38;2;11;131;121mâ•­ Redwood Upgrade Available: 1.0.0 -> 2.0.0 â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
  [38;2;11;131;121mâ”‚[39m  Checklist:                                                                    [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ”‚[39m   1. Read the release notes at: https://github.com/redwoodjs/redwood/releases  [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ”‚[39m   2. Run "yarn rw upgrade" to upgrade                                          [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m
  [38;2;11;131;121m[39m
"
`;

exports[`Upgrade is available (1.0.0-canary.100 -> 1.0.0-canary.200) Shows correct final message 1`] = `
"[32mâœ”[39m [38;2;11;131;121m[39m
  [38;2;11;131;121mâ•­ Redwood Upgrade Available: 1.0.0-canary.100 -> 1.0.0-canary.200 â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®[39m
  [38;2;11;131;121mâ”‚[39m  Checklist:                                                                    [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ”‚[39m   1. Read the release notes at: https://github.com/redwoodjs/redwood/releases  [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ”‚[39m   2. Run "yarn rw upgrade" to upgrade                                          [38;2;11;131;121mâ”‚[39m
  [38;2;11;131;121mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[39m
  [38;2;11;131;121m[39m
"
`;

exports[`Upgrade is not available (1.0.0 -> 1.0.0) Shows correct final message 1`] = `
"[32mâœ”[39m No upgrade is available
"
`;
